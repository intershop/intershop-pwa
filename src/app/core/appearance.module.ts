import { NgModule } from '@angular/core';
import { NgbPopoverConfig } from '@ng-bootstrap/ng-bootstrap';
import { ToastrModule } from 'ngx-toastr';

import { IconModule } from './icon.module';
import { ThemeService } from './utils/theme/theme.service';

@NgModule({
  imports: [
    IconModule,
    ToastrModule.forRoot({
      closeButton: true,
      timeOut: 3000,
      positionClass: 'toast-top-full-width', // toast-top-center
      preventDuplicates: true,
    }),
  ],
})
export class AppearanceModule {
  constructor(popoverConfig: NgbPopoverConfig, themeService: ThemeService) {
    popoverConfig.placement = 'top';
    popoverConfig.triggers = 'hover';
    popoverConfig.container = 'body';

    themeService.init();
  }
}
