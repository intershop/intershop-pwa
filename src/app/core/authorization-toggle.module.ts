import { ModuleWithProviders, NgModule } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';

import { AuthorizationToggleDirective } from './directives/authorization-toggle.directive';
import { AuthorizationToggleService, checkPermission } from './utils/authorization-toggle/authorization-toggle.service';
import { whenTruthy } from './utils/operators';

@NgModule({
  declarations: [AuthorizationToggleDirective],
  exports: [AuthorizationToggleDirective],
})
export class AuthorizationToggleModule {
  private static permissions = new ReplaySubject<string[]>(1);

  static forTesting(...permissions: string[]): ModuleWithProviders<AuthorizationToggleModule> {
    AuthorizationToggleModule.switchTestingPermissions(...permissions);
    return {
      ngModule: AuthorizationToggleModule,
      providers: [
        {
          provide: AuthorizationToggleService,
          useValue: {
            isAuthorizedTo: (permission: string) =>
              AuthorizationToggleModule.permissions.pipe(
                whenTruthy(),
                map(perms => checkPermission(perms, permission))
              ),
          },
        },
      ],
    };
  }

  static switchTestingPermissions(...permissions: string[]) {
    AuthorizationToggleModule.permissions.next(permissions);
  }
}

export { AuthorizationToggleService } from './utils/authorization-toggle/authorization-toggle.service';
export { AuthorizationToggleGuard } from './guards/authorization-toggle.guard';
