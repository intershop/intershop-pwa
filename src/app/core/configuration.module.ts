import { NgModule } from '@angular/core';

import { SPECIAL_HTTP_ERROR_HANDLER } from './interceptors/icm-error-mapper.interceptor';
import { createPaymentErrorHandler } from './utils/http-error/create-payment.error-handler';
import { editPasswordErrorHandler } from './utils/http-error/edit-password.error-handler';
import { LoginUserErrorHandler } from './utils/http-error/login-user.error-handler';
import { requestReminderErrorHandler } from './utils/http-error/request-reminder.error-handler';
import { updatePasswordErrorHandler } from './utils/http-error/update-password.error-handler';

@NgModule({
  providers: [
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: updatePasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useClass: LoginUserErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: requestReminderErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: editPasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: createPaymentErrorHandler, multi: true },
  ],
})
export class ConfigurationModule {}
