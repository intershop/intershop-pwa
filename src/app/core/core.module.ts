import { APP_BASE_HREF, PlatformLocation } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { ErrorHandler, NgModule, Optional, SkipSelf } from '@angular/core';
import { ServiceWorkerModule } from '@angular/service-worker';
import { TranslateModule } from '@ngx-translate/core';

import { AppearanceModule } from './appearance.module';
import { ConfigurationModule } from './configuration.module';
import { FeatureToggleModule } from './feature-toggle.module';
import { IdentityProviderModule } from './identity-provider.module';
import { ICMErrorMapperInterceptor } from './interceptors/icm-error-mapper.interceptor';
import { IdentityProviderInterceptor } from './interceptors/identity-provider.interceptor';
import { MockInterceptor } from './interceptors/mock.interceptor';
import { InternationalizationModule } from './internationalization.module';
import { StateManagementModule } from './state-management.module';
import { DefaultErrorHandler } from './utils/default-error-handler';

@NgModule({
  imports: [
    AppearanceModule,
    ConfigurationModule,
    FeatureToggleModule,
    HttpClientModule,
    IdentityProviderModule,
    InternationalizationModule,
    ServiceWorkerModule.register('ngsw-worker.js', { enabled: SERVICE_WORKER }),
    StateManagementModule,
  ],
  providers: [
    { provide: HTTP_INTERCEPTORS, useClass: ICMErrorMapperInterceptor, multi: true },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: IdentityProviderInterceptor,
      multi: true,
    },
    { provide: HTTP_INTERCEPTORS, useClass: MockInterceptor, multi: true },
    { provide: ErrorHandler, useClass: DefaultErrorHandler },
    {
      provide: APP_BASE_HREF,
      useFactory: (s: PlatformLocation, baseHref: string) => baseHref || s.getBaseHrefFromDOM(),
      deps: [PlatformLocation, [new Optional(), new SkipSelf(), APP_BASE_HREF]],
    },
  ],
  // exports needed to use the cookie banner in the AppComponent
  exports: [TranslateModule],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error('CoreModule is already loaded. Import it in the AppModule only');
    }
  }
}
