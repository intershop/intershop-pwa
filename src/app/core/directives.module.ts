import { NgModule } from '@angular/core';

import { ClickOutsideDirective } from './directives/click-outside.directive';
import { IdentityProviderCapabilityDirective } from './directives/identity-provider-capability.directive';
import { IntersectionObserverDirective } from './directives/intersection-observer.directive';
import { ProductContextAccessDirective } from './directives/product-context-access.directive';
import { ProductContextDirective } from './directives/product-context.directive';
import { ServerHtmlDirective } from './directives/server-html.directive';

@NgModule({
  declarations: [
    ClickOutsideDirective,
    IdentityProviderCapabilityDirective,
    IntersectionObserverDirective,
    ProductContextAccessDirective,
    ProductContextDirective,
    ServerHtmlDirective,
  ],
  exports: [
    ClickOutsideDirective,
    IdentityProviderCapabilityDirective,
    IntersectionObserverDirective,
    ProductContextAccessDirective,
    ProductContextDirective,
    ServerHtmlDirective,
  ],
})
export class DirectivesModule {}
