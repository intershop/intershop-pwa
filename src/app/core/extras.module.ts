import { NgModule } from '@angular/core';

import { SentryModule } from '../extensions/sentry/sentry.module';
import { SeoModule } from '../extensions/seo/seo.module';
import { TrackingModule } from '../extensions/tracking/tracking.module';

/*
 * declaration of optional modules, comment out to deactivate
 */
// tslint:disable:ng-module-sorted-fields
const importExportModules = [
  // tracking with https://tagmanager.google.com
  TrackingModule,
  // browser error tracking with https://sentry.io
  SentryModule,
  // ngx-meta integration https://www.npmjs.com/package/@ngx-meta/core
  SeoModule,
];

@NgModule({
  imports: [...importExportModules],
  exports: [...importExportModules],
})
export class ExtrasModule {}
