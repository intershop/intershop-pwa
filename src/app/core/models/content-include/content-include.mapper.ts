import { Injectable } from '@angular/core';

import { ContentConfigurationParameterMapper } from '../content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentPageletMapper } from '../content-pagelet/content-pagelet.mapper';
import { ContentPagelet } from '../content-pagelet/content-pagelet.model';

import { ContentIncludeData } from './content-include.interface';
import { ContentInclude } from './content-include.model';

@Injectable({ providedIn: 'root' })
export class ContentIncludeMapper {
  constructor(
    private contentConfigurationParameterMapper: ContentConfigurationParameterMapper,
    private contentPageletMapper: ContentPageletMapper
  ) {}

  /**
   * Converts {@link ContentIncludeData} to the model entity {@link ContentInclude} and enclosed {@link ContentPagelet}s.
   */
  fromData(data: ContentIncludeData): { include: ContentInclude; pagelets: ContentPagelet[] } {
    if (!data) {
      throw new Error('falsy input');
    }

    let pagelets: ContentPagelet[] = [];
    let pageletIDs: string[] = [];

    if (data.pagelets) {
      pageletIDs = data.pagelets.map(p => p.id);
      pagelets = data.pagelets
        .map(x => this.contentPageletMapper.fromData(x))
        .reduce((acc, val) => [...acc, ...val], []);
    }

    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);

    const include: ContentInclude = {
      id: data.link.title,
      definitionQualifiedName: data.definitionQualifiedName,
      configurationParameters,
      pageletIDs,
    };

    return { include, pagelets };
  }
}
