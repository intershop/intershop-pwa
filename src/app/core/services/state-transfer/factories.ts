import { InjectionToken } from '@angular/core';
import { StatePropertiesService } from './state-properties.service';

/**
 * The injected property contains the URL to the rest endpoint of the ICM Application.
 * It is set by {@link StatePropertiesService#getRestEndPoint}
 */
export const REST_ENDPOINT = new InjectionToken<string>('restEndpoint');

export function getRestEndPoint() {
  return (service: StatePropertiesService): string => service.getRestEndPoint();
}

/**
 * The injected property contains the base URL to the ICM Application to be used for constructing paths to static data.
 * It is set by {@link StatePropertiesService#getICMBaseURL}
 */
export const ICM_BASE_URL = new InjectionToken<string>('icmBaseURL');

export function getICMBaseURL() {
  return (service: StatePropertiesService): string => service.getICMBaseURL();
}

/**
 * The injected property contains the name of the connected ICM Site.
 * It is set by {@link StatePropertiesService#getICMApplication}
 */
export const ICM_APPLICATION = new InjectionToken<string>('icmApplication');

export function getICMApplication() {
  return (service: StatePropertiesService): string => service.getICMApplication();
}

/**
 * The injected property contains the URL of the connected ICM Server.
 * It is set by {@link StatePropertiesService#getICMServer}
 */
export const ICM_SERVER_URL = new InjectionToken<string>('icmServerURL');

export function getICMServerURL() {
  return (service: StatePropertiesService): string => service.getICMServerURL();
}
