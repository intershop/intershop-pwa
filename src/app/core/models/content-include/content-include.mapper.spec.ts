import { TestBed } from '@angular/core/testing';

import { STATIC_URL } from 'ish-core/utils/state-transfer/factories';

import { ContentIncludeData } from './content-include.interface';
import { ContentIncludeMapper } from './content-include.mapper';

describe('Content Include Mapper', () => {
  let contentIncludeMapper: ContentIncludeMapper;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [{ provide: STATIC_URL, useValue: 'http://www.example.org/static' }],
    });
    contentIncludeMapper = TestBed.get(ContentIncludeMapper);
  });

  it('should throw on falsy input', () => {
    expect(() => contentIncludeMapper.fromData(undefined)).toThrowError('falsy input');
  });

  it('should convert a complex example to complex type', () => {
    const input: ContentIncludeData = {
      link: {
        title: 'include-id',
        uri: 'uri://test',
        type: 'ContentInclude',
      },
      definitionQualifiedName: 'fq',
      displayName: 'name-include',
      pagelets: [
        {
          id: 'pagelet1',
          definitionQualifiedName: 'fq',
          displayName: 'name1',
          configurationParameters: {
            key1: {
              definitionQualifiedName: 'fq',
              value: 'hallo 1',
            },
          },
          slots: {
            slot1: {
              definitionQualifiedName: 'fq',
              pagelets: [
                {
                  id: 'pagelet11',
                  definitionQualifiedName: 'fq',
                  displayName: 'name11',
                  configurationParameters: {
                    key11: {
                      definitionQualifiedName: 'fq',
                      value: 'test',
                    },
                  },
                },
              ],
            },
          },
        },
        {
          id: 'pagelet2',
          definitionQualifiedName: 'fq',
          displayName: 'name2',
        },
      ],
    };

    const result = contentIncludeMapper.fromData(input);

    expect(result.pagelets).toHaveLength(3);
    expect(result.pagelets.map(p => p.id)).toIncludeAllMembers(['pagelet1', 'pagelet11', 'pagelet2']);
    expect(result.include).not.toBeEmpty();
    expect(result).toMatchSnapshot();
  });
});
