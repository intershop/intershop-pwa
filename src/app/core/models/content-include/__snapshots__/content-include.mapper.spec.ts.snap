// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Include Mapper should convert a complex example to complex type 1`] = `
Object {
  "include": Object {
    "configurationParameters": Object {},
    "definitionQualifiedName": "fq",
    "id": "include-id",
    "pageletIDs": Array [
      "pagelet1",
      "pagelet2",
    ],
  },
  "pagelets": Array [
    Object {
      "configurationParameters": Object {
        "key1": "hallo 1",
      },
      "definitionQualifiedName": "fq",
      "id": "pagelet1",
      "slots": Array [
        Object {
          "configurationParameters": Object {},
          "definitionQualifiedName": "fq",
          "pageletIDs": Array [
            "pagelet11",
          ],
        },
      ],
    },
    Object {
      "configurationParameters": Object {
        "key11": "test",
      },
      "definitionQualifiedName": "fq",
      "id": "pagelet11",
      "slots": Array [],
    },
    Object {
      "configurationParameters": Object {},
      "definitionQualifiedName": "fq",
      "id": "pagelet2",
      "slots": Array [],
    },
  ],
}
`;
